lib/pbench/common/logger.py

```diff
@@ -158,11 +158,13 @@ def get_pbench_logger(caller, config):
             handler = logging.FileHandler(log_dir / f"{caller}.log")
         elif config.logger_type == "devlog":
             handler = logging.handlers.SysLogHandler(address=_devlog)
+            handler.ident = f"{caller}: "
```

Comment:
Assuming that this `ident` field is the RFC `APP-NAME`, should we be decorating it with the `: `?? The context of how all these specifications fit together isn't entirely clear; but if Python is by default pulling `argv[0]` (which is what the default appears to be), and we're getting `<argv[0]>[<pid>]: ` (which we are), then wouldn't including the `: ` in `ident` result in `<ident>: [<pid>]: `? Or is the syslog handler smart enough to strip it off? Or ... ???

Anyway, right now on main, server log messages show

```
Feb 27 18:45:36 rhel8-dbutenho-hack pbench-server[69]: [2023-02-27 18:45:36 +0000] [69] [INFO] Starting gunicorn 20.1.0
```

... and I think that's exactly what we want. I infer that setting `handler.ident = caller`, if `caller == "pbench-server"`, would have exactly the same effect; but what are we actually getting with the `: `??

Reply:
To put it another way, `ident` should be a field in a message which is formatted by some downstream component; therefore, I suspect that we should not be embedding formatting (like the colon and trailing space) in the `ident` field.

Reply:
> Assuming that this `ident` field is the RFC `APP-NAME`, should we be decorating it with the `: `??

Yes.

> The context of how all these specifications fit together isn't entirely clear;

Agreed.  Never has been, and most likely never will be.  Logging is messy, like sewage plumbing.

> but if Python is by default pulling `argv[0]` (which is what the default appears to be), and we're getting `<argv[0]>[<pid>]: ` (which we are)

Where do you see in the Python standard library code or documentation that Python is setting the `SysLogHandler.ident` field?

> , then wouldn't including the `: ` in `ident` result in `<ident>: [<pid>]: `? Or is the syslog handler smart enough to strip it off? Or ... ???

The `SysLogHandler` is only responsible for sending a properly formatted [RFC 5424](https://www.rfc-editor.org/rfc/rfc5424) log message.  If the `ident` field is not set, then the RFC 5424 `APP-NAME` is not set.

> Anyway, right now on main, server log messages show
> 
> ```
> Feb 27 18:45:36 rhel8-dbutenho-hack pbench-server[69]: [2023-02-27 18:45:36 +0000] [69] [INFO] Starting gunicorn 20.1.0
> ```

This particular record is not emitted by our code but by the Gunicorn system.  Here is the output from `journalctl -f -o json-pretty` for that record:

```
{
        "MESSAGE" : "[2023-03-02 04:35:36 +0000] [320145] [INFO] Starting gunicorn 20.1.0",
        "PRIORITY" : "6",
        "SYSLOG_FACILITY" : "3",
        "SYSLOG_IDENTIFIER" : "pbench-server",
        "_AUDIT_LOGINUID" : "0",
        "_AUDIT_SESSION" : "273",
        "_BOOT_ID" : "0e62d87e7bb9409e9584bfa38d153ddd",
        "_CAP_EFFECTIVE" : "0",
        "_CMDLINE" : "/usr/bin/python3 -s /usr/bin/gunicorn --workers 16 --timeout 9000 --pid /run/pbench-server/gunicorn.pid --bind 0.0.0.0:8001 --log-syslog --log-syslog-prefix pbench-server --pythonpath /home/pbench/.local/lib/python3.9/site-packages,/opt/pbench-server/lib \"pbench.cli.server.shell:app()\"",
        "_COMM" : "gunicorn",
        "_GID" : "1000",
        "_MACHINE_ID" : "ca9bb2b1c9592a441cabb0280e957fb2",
        "_PID" : "320145",
        "_SELINUX_CONTEXT" : "system_u:system_r:container_init_t:s0:c483,c1012",
        "_STREAM_ID" : "985ecd105b3a48f89f82b4a96589db33",
        "_SYSTEMD_CGROUP" : "/system.slice/pbench-server.service",
        "_SYSTEMD_INVOCATION_ID" : "435d5de20e1b4e02a17baadadd11da54",
        "_SYSTEMD_SLICE" : "system.slice"
        "_SYSTEMD_UNIT" : "pbench-server.service",
        "_TRANSPORT" : "stdout",
        "_UID" : "1000",
        "__CURSOR" : "s=6cf94159be0a4b9e9488e0078bec9da5;i=665b1;b=0e62d87e7bb9409e9584bfa38d153ddd;m=93947019dff;t=5f5e35c2c6e41;x=9f66b2146d58413d",
        "__MONOTONIC_TIMESTAMP" : "10141609074175",
        "__REALTIME_TIMESTAMP" : "1677731736350273",
}
```

Notice that the Gunicorn system is setting the `pbench-server` as the `SYSLOG_IDENTIFIER` [1] which is the "tag" or the "APP-NAME" in RFC 5424.

[1] https://www.freedesktop.org/software/systemd/man/systemd.journal-fields.html#SYSLOG_FACILITY=

> ... and I think that's exactly what we want.

I agree, that is exactly what we want and what this change provides the log emitted by our `SysLogHandler` instance.

For example, here is the log emitted by the `shell.py` just before spawning gunicorn:

```
{
        "MESSAGE" : "2023-03-02T04:35:36.196732 INFO 320144 140126434179968 pbench-shell shell run_gunicorn 200 -- Starting Gunicorn Pbench Server application",
        "PRIORITY" : "6",
        "SYSLOG_FACILITY" : "1",
        "SYSLOG_RAW" : [...],
        "_AUDIT_LOGINUID" : "0",
        "_AUDIT_SESSION" : "273",
        "_BOOT_ID" : "0e62d87e7bb9409e9584bfa38d153ddd",
        "_CAP_EFFECTIVE" : "0",
        "_CMDLINE" : "/usr/bin/python3 /opt/pbench-server/bin/pbench-server",
        "_COMM" : "pbench-server",
        "_GID" : "1000",
        "_HOSTNAME" : "intlab-002.ctrl.perf-infra.lab.eng.rdu2.redhat.com",
        "_MACHINE_ID" : "ca9bb2b1c9592a441cabb0280e957fb2",
        "_PID" : "320144",
        "_SELINUX_CONTEXT" : "system_u:system_r:container_init_t:s0:c483,c1012",
        "_SOURCE_REALTIME_TIMESTAMP" : "1677731736196789",
        "_SYSTEMD_CGROUP" : "/system.slice/pbench-server.service"
        "_SYSTEMD_INVOCATION_ID" : "435d5de20e1b4e02a17baadadd11da54",
        "_SYSTEMD_SLICE" : "system.slice",
        "_SYSTEMD_UNIT" : "pbench-server.service",
        "_TRANSPORT" : "syslog",
        "_UID" : "1000",
        "__CURSOR" : "s=6cf94159be0a4b9e9488e0078bec9da5;i=665b0;b=0e62d87e7bb9409e9584bfa38d153ddd;m=93946ff4677;t=5f5e35c2a16ba;x=65153f50a489cd23",
        "__MONOTONIC_TIMESTAMP" : "10141608920695",
        "__REALTIME_TIMESTAMP" : "1677731736196794",
}
```

Notice that there is no `SYSLOG_IDENTIFIER` field recorded for that record, and so `journalctl` is using the `_COMM` field instead, which ends up being what we want in that case.

But then we start registering end points, and we see:

```
Mar 02 04:45:35 example.com gunicorn[320386]: 2023-03-02T04:45:35.340051 INFO 320386 140256262540160 pbench.server.api __init__ register_endpoints 58 -- Registering service endpoints with base URI /api/v1
```

And our log entry now has a generic `gunicorn` value instead of the expected `pbench.server.api`, the logger name.

So the `pbench-server` name happens to work on the parent process of the service, but not for the spawned workers.

> I infer that setting `handler.ident = caller`, if `caller == "pbench-server"`, would have exactly the same effect; but what are we actually getting with the `: `??

Would you mind explaining how you inferred that based on what you are seeing?

I'll post what I see in the logs when I use `jenkins/runlocal` with this change.

Reply:
Here is what it looks like now:

```
Mar 02 04:59:16 gandalf..com pbench-shell[45]: 2023-03-02T04:59:16.292052 INFO 45 139953109665600 pbench-shell logger log 44 -- Waiting at most 120 seconds for database instance postgresql://pbenchcontainer:pbench@localhost:5432/pbenchcontainer to become available.
Mar 02 04:59:16 gandalf..com pbench-shell[45]: 2023-03-02T04:59:16.292525 INFO 45 139953109665600 pbench-shell logger log 44 -- Waiting at most 120 seconds for the Elasticsearch instance http://localhost:9200 to become available.
Mar 02 04:59:16 gandalf..com pbench-shell[45]: 2023-03-02T04:59:16.292820 WARNING 45 139953109665600 pbench-shell logger log 44 -- OpenID Connect client not configured,
Mar 02 04:59:16 gandalf..com pbench-shell[45]: 2023-03-02T04:59:16.292865 INFO 45 139953109665600 pbench-shell logger log 44 -- Performing database setup
Mar 02 04:59:16 gandalf..com pbench-shell[45]: 2023-03-02T04:59:16.418741 INFO 45 139953109665600 pbench-shell logger log 44 -- Performing Elasticsearch indexing setup
```

Now we have set the program name to `pbench-shell`.

Reply:
Well, the result seems to be right; although if that's really how Python's `ident` attribute is supposed to work (requiring the `: `), that sounds flaky to me.

From the Python documentation:

> Changed in version 3.3: (See: [bpo-12419](https://bugs.python.org/issue?@action=redirect&bpo=12419).) In earlier versions, there was no facility for an “ident” or “tag” prefix to identify the source of the message. This can now be specified using a class-level attribute, defaulting to "" to preserve existing behaviour, but which can be overridden on a SysLogHandler instance in order for that instance to prepend the ident to every message handled. Note that the provided ident must be text, not bytes, and is prepended to the message exactly as is.

Following through the addition of `ident` to the Python handler, we're redirected to `syslog.openlog`, which says:

> ident is an arbitrary identification string which future syslog invocations will prefix to each message. This is intended to identify the source of the message, and people conventionally set it to the name of the program that will submit the messages.
> 
> If ident is NULL, or if openlog is not called, the default identification string used in Syslog messages will be the program name, taken from argv[0].

Neither source (nor any others I found) even suggest that the string should include any punctuation. Have you tried without it? This all just seems weird; but since it seems to work, I'll withdraw my objection and simply admit the world is full of squirrels and makes no sense...

Reply:
> Neither source (nor any others I found) even suggest that the string should include any punctuation. Have you tried without it?

See PR #3298.